src/linker/Linker.Steps/RemoveUnreachableBlocksStep.cs

```diff
@@ -16,43 +16,268 @@ namespace Mono.Linker.Steps
+		void AddMethodForProcessing (MethodDefinition method)
+		{
+			Debug.Assert (!processedMethods.ContainsKey (method));
+
+			var processingNode = new ProcessingNode (method, -1);
+
+			var stackNode = processingStack.AddFirst (processingNode);
+			processingMethods.Add (method, stackNode);
+			processingStackVersion++;
+		}
+
+		void StoreMethodAsProcessedAndRemoveFromQueue (LinkedListNode<ProcessingNode> stackNode, Instruction methodValue)
+		{
+			Debug.Assert (stackNode.List == processingStack);
+			Debug.Assert (methodValue != null);
+
+			var method = stackNode.Value.Method;
+			processingMethods.Remove (method);
+			processingStack.Remove (stackNode);
+			processingStackVersion++;
+
+			processedMethods[method] = methodValue;
 		}
 
-		bool TryGetConstantResultInstructionForMethod (MethodDefinition method, out Instruction constantResultInstruction)
+		void ProcessStack ()
 		{
-			if (constExprMethods.TryGetValue (method, out constantResultInstruction))
-				return constantResultInstruction != null;
+			while (processingStack.Count > 0) {
+				var stackNode = processingStack.First;
+				var method = stackNode.Value.Method;
+
+				bool treatUnprocessedDependenciesAsNonConst = false;
+				if (stackNode.Value.LastAttemptStackVersion == processingStackVersion) {
+					// Loop was detected - the stack hasn't changed since the last time we tried to process this method
+					// as such there's no way to resolve the situation (running the code below would produce the exact same result).
+
+					// Observation:
+					//   All nodes on the stack which have `LastAttemptStackVersion` equal to `processingStackVersion` are part of the loop
+					//   meaning removing any of them should break the loop and allow to make progress.
+					//   There might be other methods in between these which don't have the current version but are dependencies of some of the method
+					//   in the loop.
+					//   If we don't process these, then we might miss constants and branches to remove. See the doc
+					//   `constant-propagation-and-branch-removal.md` in this repo for more details and a sample.
+
+					// To fix this go over the stack and find the "oldest" node with the current version - the "oldest" node which
+					// is part of the loop:
+					var lastNodeWithCurrentVersion = stackNode;
+					for (var currentNode = stackNode; currentNode != null; currentNode = currentNode.Next) {
+						if (currentNode.Value.LastAttemptStackVersion == processingStackVersion)
+							lastNodeWithCurrentVersion = currentNode;
+					}
+
+					// There should be at least 2 nodes with the latest version to form a loop
+					Debug.Assert (lastNodeWithCurrentVersion != stackNode);
+
+					// Now go back over all nodes from the "oldest" one back to the top and find any nodes which are not of current version.
+					// For all of them, move them to the top of the stack.
+					var candidateNodeToMoveToTop = lastNodeWithCurrentVersion;
+					bool foundNodesWithNonCurrentVersion = false;
```

Comment:
You could also compute foundNodesWithNonCurrentVersion in the loop above to avoid walking the loop nodes twice in case there are none.

Reply:
Actually it can all be done in one backward full walk of the stack - changed the implementation.